### ISSUE TITLE
php7.2-fpm fails to start on reboot

### ISSUE THREAD

**Issue by leewillis77** [oerdnj/deb.sury.org#833]
Hi; 

I recently installed php7.2 on a DigitalOcean VM running Ubuntu 16.04.4. The following packages are installed:

```
php7.2-cli                       7.2.3-1+ubuntu16.04.1+deb.sury.org+1
php7.2-common                    7.2.3-1+ubuntu16.04.1+deb.sury.org+1
php7.2-curl                      7.2.3-1+ubuntu16.04.1+deb.sury.org+1
php7.2-fpm                       7.2.3-1+ubuntu16.04.1+deb.sury.org+1
```

The service starts/stops normally via `service php7.2-fpm stop/start` - however it fails to start properly on a fresh reboot, with the following logged:

```
Mar 13 18:36:08  systemd[1]: Starting The PHP 7.2 FastCGI Process Manager...
Mar 13 18:37:39  systemd[1]: php7.2-fpm.service: Start operation timed out. Terminating.
Mar 13 18:37:39  systemd[1]: Failed to start The PHP 7.2 FastCGI Process Manager.
Mar 13 18:37:39  systemd[1]: php7.2-fpm.service: Unit entered failed state.
Mar 13 18:37:39  systemd[1]: php7.2-fpm.service: Failed with result 'timeout'.
```

I suspect that the underlying issue is [this](https://github.com/jedisct1/libsodium-php/issues/94) - however it doesn't seem (that I can tell) that there's a workaround to that. However - that leaves this PPA unusable in the meantime. 

Adding the following to the `[Service]` section of `etc/systemd/system/multi-user.target.wants/php7.2-fpm.service` means that restarts will be re-attempted every three seconds until /dev/random is unblocked and libsodium can start without blocking.  
```
Restart=always
RestartSec=3 
```

Can / should this be added to the repo - or is there a better solution I'm missing here?

**Comment by oerdnj** [comment]
Adding these lines as a default would be very wrong. You don't want the service to always restart. The default is fine.

Just save the random seed on shutdown and reseed on boot. This won't really help with fresh instances, but it would help with reboots.

**Comment by oerdnj** [comment]
In fact, this is already done by either `/etc/init.d/urandom` or `/lib/systemd/system/urandom.service`, so your underlying problem is probably something else.

**Comment by leewillis77** [comment]
Hi; 

Thanks for the feedback.

**Comment by acalimit** [comment]
Same problem on fresh Debian8.10. Failed with result 'timeout'.
I got this result on three different kvm VPSs. Php7.1-fpm does not have this problem.

But it's working good on local virtualbox machine. Very weird.

**Comment by schlegel11** [comment]
Same problem here too.
Failed to start or better timed out after kernel upgrade on a digital ocean debian instance.
Adding the lines in the first post solves the issue (anyway the service startup takes roundabout 2 minutes).
Is this really a final solution?

**Comment by MatthiasKuehneEllerhold** [comment]
It works under debian stretch... Is there something ubuntu specific that prevent it from starting?

**Comment by Kulgar** [comment]
You should take a look at this conversation: https://github.com/jedisct1/libsodium-php/issues/94

I think it is rather a "OS hosted on cloud" problem, according to what I read in this other post.

**Comment by MatthiasKuehneEllerhold** [comment]
Ok, our servers are not in the cloud. That could be it.

**Comment by MACscr** [comment]
im confused, this problem is closed, yet still seems to be an issue. Whats the final and official solution and why doesnt work out of the box on boot when enabled? Shouldnt "systemctl enable php7.2-fpm" be enough?

**Comment by todeveni** [comment]
> Whats the final and official solution and why doesnt work out of the box on boot when enabled?

Though not offical, the problem lies in your cloud provider and your server not having enough random bytes for libsodium/php to start in the allotted time (hence the timeout in error log) and [here's one solution](https://github.com/jedisct1/libsodium-php/issues/94#issuecomment-405598859) to fix this by installing `haveged`.

**Comment by MACscr** [comment]
I’m confused as I don’t use libsodium or any non-native packages.

- Mark

**Comment by oerdnj** [comment]
Have you tried reading documentation? I’ve heard it helps with confusion...

**Comment by todeveni** [comment]
> I’m confused as I don’t use libsodium or any non-native packages.

Sodium is part of PHP core as of PHP 7.2 and `php7.2-fpm` depends on `libsodium`.
